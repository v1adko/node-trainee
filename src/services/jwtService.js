import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
