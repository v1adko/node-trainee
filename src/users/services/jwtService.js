const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
