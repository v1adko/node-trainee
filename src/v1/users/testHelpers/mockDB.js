import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
