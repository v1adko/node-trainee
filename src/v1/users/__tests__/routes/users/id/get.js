import simulate from '../../../../../../utils/tests/requestHelper';
import jwtService from '../../../../../../services/jwtService';
import User from '../../../../models/user';
import mockDB from '../../../../testHelpers/mockDB';
import UserFields from '../../../../../../utils/tests/testUserFields';

const ROUTE = '/v1/users';

const {
  username, password, invalidUserId, invalidToken
} = new UserFields(
  ROUTE
);

let user;
let userToken;

async function clean() {
  await mockDB.cleanDB();
  userToken = null;
}

async function createUser() {
  const userActor = new User();
  userToken = jwtService.generateJwt(userActor);
  user = await mockDB.createUser(username, password);
}

beforeAll(async () => {
  await mockDB.createDefaultUsers();
});

describe(`Test the ${ROUTE}/:id path`, () => {
  afterAll(clean);
  beforeAll(createUser);

  it('should return user by id in response on GET method', async () => {
    const route = `${ROUTE}/${user.id}`;
    const result = await simulate.get(route, 200, userToken);
    const { id, username: name, role } = result.body;

    expect(id).toBe(user.id.toString());
    expect(name).toBe(user.username);
    expect(role).toBe(user.role);
  });

  it('should not return user in response on GET method, because user token is not valid', async () => {
    const route = `${ROUTE}/${user.id}`;
    const result = await simulate.get(route, 401, invalidToken);
    const { auth, message } = result.body;

    expect(auth).toBe(false);
    expect(message).toBe('Invalid token, please repeat authentication.');
  });

  it('should not return user in response on GET method, because user id is invalid', async () => {
    const route = `${ROUTE}/${invalidUserId}`;
    const result = await simulate.get(route, 400, userToken);
    const { message } = result.body;

    expect(message).toBe('User id is invalid');
  });
});
