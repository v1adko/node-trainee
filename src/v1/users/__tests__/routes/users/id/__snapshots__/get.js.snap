// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user id is invalid 1`] = `"\\"id\\" must only contain hexadecimal characters"`;

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user token is not valid 1`] = `"Invalid token, please repeat authentication."`;
